test/unit/ngsiv2/expressions/expressionBasedTransformations-test.js

```diff
@@ -0,0 +1,774 @@
+        // Case: Update for a Float attribute without expressions
+
+        var values = [
+            {
+                name: 'e',
+                type: 'Float',
+                value: 0.44
+            }
+        ];
+
+        beforeEach(function() {
+            nock.cleanAll();
+
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', 'gardens')
+                .post('/v2/entities/light1/attrs', utils.readExampleFile(
+                    './test/unit/ngsiv2/examples/contextRequests/updateContextExpressionPlugin12.json'))
+                .reply(204);
+        });
+
+        it('should apply the expression before sending the values', function(done) {
+            iotAgentLib.update('light1', 'Light', '', values, function(error) {
+                should.not.exist(error);
+                contextBrokerMock.done();
+                done();
+            });
+        });
+    });
+
+    describe('When an update comes for attributes with numeric expressions and type float', function() {
+        // Case: Update for a Float attribute with arithmetic expression
+
+        var values = [
+            {
+                name: 'e',
+                type: 'Float',
+                value: 0.44
+            }
+        ];
+
+        beforeEach(function() {
+            nock.cleanAll();
+
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', 'gardens')
+                .post('/v2/entities/ws1/attrs', utils.readExampleFile(
+                    './test/unit/ngsiv2/examples/contextRequests/updateContextExpressionPlugin8.json'))
+                .reply(204);
+        });
+
+        it('should apply the expression before sending the values', function(done) {
+            iotAgentLib.update('ws1', 'WeatherStation', '', values, function(error) {
+                should.not.exist(error);
+                contextBrokerMock.done();
+                done();
+            });
+        });
+    });
+
+    describe('When an update comes for attributes with string expressions and type float', function() {
+        // Case: Update for a Float attribute with string expression
+
+        var values = [
+            {
+                name: 'e',
+                type: 'Float',
+                value: 0.44
+            }
+        ];
+
+        beforeEach(function() {
+            nock.cleanAll();
+
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', 'gardens')
+                .post('/v2/entities/ws1/attrs', utils.readExampleFile(
+                    './test/unit/ngsiv2/examples/contextRequests/updateContextExpressionPlugin12.json'))
```

Comment:
The definition for e in WeatherStationMultiple is:

```
{
                        object_id: 'e',
                        name: 'consumption',
                        type: 'Float',
                        expression: '${trim(@spaces)}'
}
```

If there isn't any `spaces` attribute, then `${trim(@spaces)}` results in `""` empty string. However, updateContextExpressionPlugin12.json shows:

```
  "consumption": {
    "type": "Float",
    "value": 0.44
  }
```

It is like the expression were being ignored and the attribute value passed directly to CB...

Reply:
Fixed in 4026d75714dd6ef51230bd01c8482c0a89e36923

Reply:
Considering the code previous to the fix, I understand that `${trim(@spaces)}` should evaluate to `""` (as `spaces` attributes doesn't exist), which cast to Float I understand is `0.0`. Thus, why we had (previous to the fix)?

```
  "consumption": {
    "type": "Float",
    "value": 0.44
  }
```

instead to

```
"consumption": {
    "type": "Float",
    "value": 0.0
  }
```

Although this has been fixed changing expression `'${trim(@updated)}'` it would be great to know how it worked before. That behaviour could hide some kind of bug...

Reply:
This worked because the plugin checks if attribute name matches the expression variables before applying it.

In 179 of[expressionParser.js](https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/expressionParser.js):
* list argument contains the active attributes provisioned for the device. These attributes may include expressions.
* context contains the measure received from the Southbound, e.g., {"consumption":0.44}
* contextAvailable function (line 170) extracts the variables from the expression and returns true if they match the measure. In this example, the comparison of line 176 is between ["spaces"] and ["consumption"]. Therefore, the expression is not applied and the IOTA will send to the CB an updateContext query that contains directly the NGSI translation of the Southbound measure.

Reply:
Ok, now I understand how it works more or less. If that behaviour is not explained in the documentation (I haven't find it, but I could be wrong) I'd suggest to add the following explanatory text to expressionLanguage.md (maybe at then end of "Expression execution" section):

> Note that expressions are only applied if the attribute name (as recevied by the IOT Agent in the southbound interface) matches the expression variable. Otherwise the southbound value is used directly. Let's illustrate with the following example:
> ```
>   "consumption": {
>    "type": "Float",
>    "value": "${trim(@spaces)}"
>  }
> ````
>
> Case 1: the following measure is received at southbound interface:
> ```
> consumption: 0.44
> ```
> as `spaces` attribute is not included, then the expression is not applied and the
`consumption` measure value is directly used, so the following is sent to CB:
> ``` 
>"consumption": {
>  "type": "Float",
>  "value": 0.44
> }
> ```
> Case 2: the following measure is recieved at southbound interface:
>
> ```
> consumption: 0.44
> spaces: "  foobar  "
>```
> as `spaces` is included, then the expression is evaluated, so overriding the 0.44 value and sending the following to CB:
>
> ``` 
>"consumption": {
>  "type": "Float",
>  "value": "foobar"
> }
> ```

Please, ensure case 2 is as I describe. I could be wrong. Maybe it is a good idea to include a test case to ensure it.

Reply:
I have added the example you proposed with just one minor change:
```
  "consumption": {
   "type": "String",
   "value": "${trim(@spaces)}"
 }
````

As you can see I have used String type instead of Float. If we used a Float type, at the end "foobar" would be cast to Float, getting null value.

Reply:
In my eyes, it is not necessary to add new unit tests. These cases are already covered in several of the already existing tests, see the lines 232, 267 and 303.

Reply:
As far as I understand, the test cases you refer (L232, L267 and L303) use a pattern in which the expression includes the attribute or no expression is used, e.g:

```
{
                        object_id: 'p',
                        name: 'pressure',
                        type: 'Integer',
                        expression: '${@pressure * / 20}'
},
...
{
                        object_id: 'h',
                        name: 'humidity',
                        type: 'Percentage'
},
...
```

What I mean is a pattern in which the expression includes a *different* attribute. E.g.:

{
                        object_id: 'x',
                        name: 'consumption_x',
                        type: 'Float',
                        expression: '${@pressure * 20}'
},

Case 1: as test input use:

```

        var values = [
            {
                name: 'x',
                type: 'Float',
                value: 0.44
            }
];
```

And get as output:

```
"consumption_X": {
   "type": "Float",
   "value": 0.44
 }
```

Case 2: as test input use:

```
        var values = [
            {
                name: 'p'',
                type: 'Integer',
                value: 10
            }
];
```

And get as output (expression overrides "direct" value):

```
"consumption_x": {
   "type": "Float",
   "value": 200
 }
```

Case 3: as test input use:

```
        var values = [
            {
                name: 'x',
                type: 'Float',
                value: 0.44
            },
            {
                name: 'p'',
                type: 'Integer',
                value: 10
            }
];
```
(Same as case 2, but with an "overriding" situation)

And get as output:

```
"consumption_x": {
   "type": "Float",
   "value": 200
 }
```

What do you think?

Reply:
Ok, now I get your point! I will add these new tests. :)

Reply:
Fixed in e3bc553d7598e699abe68f22bc4da9bbb3c6cf85